Source/Core/Core/USBUtils.cpp

```diff
@@ -0,0 +1,303 @@
+  if (dev_info == INVALID_HANDLE_VALUE)
+    return {};
+  SP_DEVINFO_DATA dev_info_data;
+  dev_info_data.cbSize = sizeof(SP_DEVINFO_DATA);
+
+  for (DWORD i = 0; SetupDiEnumDeviceInfo(dev_info, i, &dev_info_data); ++i)
+  {
+    TCHAR instance_id[MAX_DEVICE_ID_LEN];
+    if (CM_Get_Device_ID(dev_info_data.DevInst, instance_id, MAX_DEVICE_ID_LEN, 0) != CR_SUCCESS)
+      continue;
+
+    const std::wstring_view id_wstr(instance_id);
+    if (id_wstr.find(filter) == std::wstring::npos)
+      continue;
+
+    std::wstring property_value =
+        Common::GetDeviceProperty(dev_info, &dev_info_data, &DEVPKEY_Device_FriendlyName);
+    if (property_value.empty())
+    {
+      property_value =
+          Common::GetDeviceProperty(dev_info, &dev_info_data, &DEVPKEY_Device_DeviceDesc);
+    }
+
+    if (!property_value.empty())
+      device_name = WStringToUTF8(property_value);
+    break;
+  }
+
+  SetupDiDestroyDeviceInfoList(dev_info);
+  return device_name;
+}
+#endif
+
+#ifdef HAVE_LIBUDEV
+static std::optional<std::string> GetDeviceNameUsingHWDB(u16 vid, u16 pid)
+{
+  std::optional<std::string> device_name;
+  udev* udev = udev_new();
+  if (!udev)
+    return {};
+
+  Common::ScopeGuard udev_guard{[&] { udev_unref(udev); }};
+
+  udev_hwdb* hwdb = udev_hwdb_new(udev);
+  if (!hwdb)
+    return {};
+
+  Common::ScopeGuard hwdb_guard{[&] { udev_hwdb_unref(hwdb); }};
+
+  const std::string modalias = fmt::format("usb:v{:04X}p{:04X}*", vid, pid);
+  udev_list_entry* entries = udev_hwdb_get_properties_list_entry(hwdb, modalias.c_str(), 0);
+  if (!entries)
+    return {};
+
+  udev_list_entry* device_name_entry =
+      udev_list_entry_get_by_name(entries, "ID_MODEL_FROM_DATABASE");
+  if (!device_name_entry)
+    return {};
+
+  device_name = udev_list_entry_get_value(device_name_entry);
+
+  return device_name;
+}
+#endif
+
+// libusb can cause BSODs with certain bad OEM drivers on Windows when opening a device.
+// All offending drivers are known to depend on the BthUsb.sys driver, which is a Miniport Driver
+// for Bluetooth.
+// Known offenders:
+// - btfilter.sys from Qualcomm Atheros Communications
+// - ibtusb.sys from Intel Corporation
+#if defined(__LIBUSB__) && !defined(_WIN32)
+static std::optional<std::string> GetDeviceNameUsingLibUSB(u16 vid, u16 pid)
+{
+  std::optional<std::string> device_name;
+  LibusbUtils::Context context;
+
+  if (!context.IsValid())
+    return {};
+
+  context.GetDeviceList([&device_name, vid, pid](libusb_device* device) {
+    libusb_device_descriptor desc;
+
+    if (libusb_get_device_descriptor(device, &desc) != LIBUSB_SUCCESS)
+      return true;
+
+    if (desc.idVendor == vid && desc.idProduct == pid)
+    {
+      libusb_device_handle* handle;
+      if (libusb_open(device, &handle) == LIBUSB_SUCCESS)
+      {
+        unsigned char buffer[256];
+        if (desc.iProduct &&
+            libusb_get_string_descriptor_ascii(handle, desc.iProduct, buffer, sizeof(buffer)) > 0)
```

Comment:
Sorry, this needs a rebase now and this can be changed to `LibusbUtils::GetStringDescriptor`.